adr/keycloak-adr.md

```diff
@@ -1,264 +1,139 @@
-- [OAuth2 and OpenID Connect](https://www.keycloak.org/docs/latest/server_admin/index.html#_oauth2)
-#### Additional References for GitLab and OAuth Authentication
-- **GitLab and OAuth2 Integration**: For a comprehensive guide on integrating GitLab with OAuth2 for authentication, refer to the [GitLab OAuth2 provider documentation](https://docs.gitlab.com/ee/integration/oauth_provider.html). This resource provides detailed instructions on setting up GitLab as an OAuth2 provider, enabling secure and efficient user authentication.
-- **Keycloak and OAuth2**: To understand how Keycloak can be configured as an OAuth2 authentication provider for various services, including GitLab, visit the [Keycloak OAuth2 documentation](https://www.keycloak.org/docs/latest/securing_apps/index.html#_oauth2). This documentation offers insights into the OAuth2 protocol, how Keycloak supports it, and the steps required to integrate Keycloak with services requiring OAuth2 authentication.
-
-#### Additional References for Keycloak Custom Event Listener
-For detailed guidance on creating and deploying a custom event listener in Keycloak, refer to the [Keycloak Custom Event Listener documentation](https://www.keycloak.org/docs/latest/server_development/index.html#_events). This resource provides comprehensive instructions on how to extend Keycloak's functionality through the development of custom event listeners, enabling tailored handling of authentication and user management events.
-
-#### The Update Framework (TUF) and go-tuf
-
-- **go-tuf (The Update Framework in Go)**: For insights into secure distribution and integrity verification mechanisms, go-tuf implements The Update Framework (TUF) in Go. TUF is designed to prevent various attacks on software update systems and can inspire approaches to ensuring the integrity of immutable logs with hash chaining. More information can be found in the [go-tuf GitHub repository](https://github.com/theupdateframework/go-tuf).
-
-#### nats.io references
-- [NATS Documentation](https://docs.nats.io/)
-- [NATS on Kubernetes](https://docs.nats.io/nats-on-kubernetes/nats-on-kubernetes)
-- [NATS Security](https://docs.nats.io/nats-on-kubernetes/nats-on-kubernetes)
+#### SAML Integration
+**Pros:**
+- Lets you automatically sync user groups between Keycloak and GitLab, making access control more accessible.
+
+**Cons:**
+- Setting it up can be complicated, so careful alignment between Keycloak and GitLab is needed.
+- To use all features, you might need a higher tier of GitLab, which could mean higher costs.
+- If you have a lot of users, you might need to pay for more expensive versions of Keycloak, which could be a bit of a hassle.
 
+### Provision Users
+Keycloak will provision users and manage user roles, groups, and permissions. This ensures user access is managed centrally and consistently across all applications and services.
 
+#### SCIM Integration
```

Comment:
Add **Automatic provisioning on first SSO login** to **Provision Users**. This is actually the most common case and only has one minor downside. I'd probably actually use this option for GitLab initially.

```suggestion
#### Automatic provisioning on first SSO login
**Pros:**
- Common standard SSO integration method
- The new GitLab user account is created on demand, no provisioning delay for the new user to login.

**Cons:**
- New user accounts don't exist in GitLab before the new user's first login. This prevents other users in GitLab from interacting with the new user (assigning tasks, @ mentions, etc) before their first login.

#### SCIM Integration
```